libproxy/proxy-private.cpp

```diff
@@ -0,0 +1,488 @@
+
+		/* If a PAC already exists, but came from a different URL than the one specified, remove it */
+		if (this->pac) {
+			if (this->pacurl->to_string() != confurl.to_string()) {
+				delete this->pacurl;
+				delete this->pac;
+				this->pacurl = NULL;
+				this->pac    = NULL;
+			}
+		}
+
+		/* Try to load the PAC if it is not already loaded */
+		if (!this->pac) {
+			this->pacurl = new url(confurl);
+			this->pac = confurl.get_pac();
+			if (debug) {
+				if (!this->pac)
+					cerr << "Unable to download PAC!" << endl;
+				else
+					cerr << "PAC received!" << endl;
+			}
+		}
+	}
+
+	return rtv;
+}
+
+void proxy_factory::run_pac(url &realurl, const url &confurl, vector<string> &response) {
+	/* In case of either PAC or WPAD, we'll run the PAC */
+	if (this->pac && (confurl.get_scheme() == "wpad" || confurl.get_scheme().substr(0, 4) == "pac+") ) {
+		vector<pacrunner_extension*> pacrunners = this->mm.get_extensions<pacrunner_extension>();
+
+		/* No PAC runner found, fall back to direct */
+		if (pacrunners.size() == 0) {
+			if (debug) cerr << "Unable to find a required pacrunner!" << endl;
+			return;
+		}
+
+		/* Run the PAC, but only try one PACRunner */
+		if (debug) cerr << "Using pacrunner: " << typeid(*pacrunners[0]).name() << endl;
+		string pacresp = pacrunners[0]->get(this->pac, this->pacurl->to_string())->run(realurl);
+		if (debug) cerr << "Pacrunner returned: " << pacresp << endl;
+		format_pac_response(pacresp, response);
+	}
+}
+
+void proxy_factory::clear_cache() {
+	this->wpad = false;
+	if (this->pac)    { delete this->pac;    this->pac = NULL; }
+	if (this->pacurl) { delete this->pacurl; this->pacurl = NULL; }
+}
+
+void proxy_factory::lock() {
+#ifdef WIN32
+	WaitForSingleObject(this->mutex, INFINITE);
+#else
+	pthread_mutex_lock(&this->mutex);
+#endif
+}
+
+void proxy_factory::unlock() {
+#ifdef WIN32
+	ReleaseMutex(this->mutex);
+#else
+	pthread_mutex_unlock(&this->mutex);
+#endif
+}
+
+};
+
+struct pxProxyFactory_ {
+	libproxy::proxy_factory pf;
+};
+
+extern "C" DLL_PUBLIC struct pxProxyFactory_ *px_proxy_factory_new(void) {
+	return new struct pxProxyFactory_;
+}
+
+extern "C" DLL_PUBLIC char** px_proxy_factory_get_proxies(struct pxProxyFactory_ *self, const char *url) {
+	std::vector<std::string> proxies;
+	char** retval;
+
+	// Call the main method
+	try {
+		proxies = self->pf.get_proxies(url);
+		retval  = (char**) malloc(sizeof(char*) * (proxies.size() + 1));
+		if (!retval) return NULL;
+	} catch (std::exception&) {
+		// Return NULL on any exception
+		return NULL;
+	}
+
+	// Copy the results into an array
+	// Return NULL on memory allocation failure
+	retval[proxies.size()] = NULL;
+	for (size_t i=0 ; i < proxies.size() ; i++) {
+		retval[i] = strdup(proxies[i].c_str());
+		if (retval[i] == NULL) {
+			for (int j=i-1 ; j >= 0 ; j--)
+				free(retval[j]);
+			free(retval);
+			return NULL;
+		}
+	}
+	return retval;
+}
+
+extern "C" DLL_PUBLIC void px_proxy_factory_free(struct pxProxyFactory_ *self) {
```

Comment:
Hmm. actually - Windows; do we even have dbus capabilities there?!?!

Reply:
I don't know.

Reply:
DBus on windows is possible (just without fd passing), but unusual and rare. So for applications/libraries which require libproxy on windows, this would probably be a burden.